### PR TITLE
Adds capability to remember recent hosts.

### PR DESCRIPTION
**Pull request by stratomda**
Adds capability to remember recent hosts. When entering the Master URI a list of recent Master URIs is presented. This implements the Remember recent hosts task in rosjava/android_core#257.
- [X] Remember recent hosts (useful when changing masters)

### CODE REVIEW

File: android_10/src/org/ros/android/MasterChooser.java [hunk 1]
```diff
@@ -348,6 +372,50 @@ protected boolean isQRCodeReaderInstalled(Intent intent) {
   }
 
   /**
+   * Stores the recent Master URI to preferences. This implementation does not use
```
Comment (adamantivm) [inline android_10/src/org/ros/android/MasterChooser.java]: nit: rewording. Perhaps something like: "adds the given URI to the list of recent Master URIs shared preferences". And perhaps the method name should be `addRecentMasterURI` instead.

File: android_10/src/org/ros/android/MasterChooser.java [hunk 2]
```diff
@@ -348,6 +372,50 @@ protected boolean isQRCodeReaderInstalled(Intent intent) {
...
+   * {@link android.content.SharedPreferences.Editor#putStringSet(String, Set)}
+   * since it is not available in API 10.
+   * @param uri Master URI string to store.
+   */
+  private void setRecentMasterURI(String uri)  {
```
Comment (adamantivm) [inline android_10/src/org/ros/android/MasterChooser.java]: please review formatting. For example: open should go on the same line as the control statement with one blank space (not zero, not two), etc.
Hint: it is possible to configure Android Studio to do this for you.

File: android_10/src/org/ros/android/MasterChooser.java [hunk 3]
```diff
@@ -348,6 +372,50 @@ protected boolean isQRCodeReaderInstalled(Intent intent) {
...
+    List<String> recentURIs = getRecentMasterURIs();
+    if(!recentURIs.contains(uri)){
+      recentURIs.add(0, uri);
+      if(recentURIs.size() > RECENT_MASTER_HISTORY_COUNT)
+        recentURIs = recentURIs.subList(0, RECENT_MASTER_HISTORY_COUNT);
+    }
+
+    SharedPreferences.Editor editor = getPreferences(MODE_PRIVATE).edit();
+    editor.putString(PREFS_KEY_NAME, uri);
+    for(int i=0; i<recentURIs.size(); i++)
```
Comment (adamantivm) [inline android_10/src/org/ros/android/MasterChooser.java]: i.e.: here, bring open brace to the line with for

**Review by adamantivm — CHANGES_REQUESTED** [review]
Great contribution @stratomda !! I left only a few minor style / name nits if you could please address.

**Comment by jubeira** [comment]
The code is looking good @stratomda !
Apart from the nits that @adamantivm pointed out, I would leave a space after statements like `for`, `if`, etc. so as to match the general coding style of rosjava.
In summary:
- Braces in the same line of statements, with one space after the end of the control statement.
- Parenthesis with one space after the control statements.

I haven't tried the changes in an app yet, I will do soon when I have some time. Thanks again for your work!

**Comment by stratomda** [comment]
Sorry about those formatting issues. I'll make sure to follow those on the next commit!

File: android_10/src/org/ros/android/MasterChooser.java [hunk 4]
```diff
@@ -348,6 +372,52 @@ protected boolean isQRCodeReaderInstalled(Intent intent) {
   }
 
   /**
+   * Adds the given URI to the list of recent Master URIs stored in shared preferences.
+   * This implementation does not use
+   * {@link android.content.SharedPreferences.Editor#putStringSet(String, Set)}
+   * since it is not available in API 10.
+   * @param uri Master URI string to store.
+   */
+  private void setRecentMasterURI(String uri) {
```
Comment (adamantivm) [inline android_10/src/org/ros/android/MasterChooser.java]: how about renaming this to `addRecentMaterURI` ?

File: android_10/src/org/ros/android/MasterChooser.java [hunk 5]
```diff
@@ -348,6 +372,52 @@ protected boolean isQRCodeReaderInstalled(Intent intent) {
...
+    List<String> recentURIs = getRecentMasterURIs();
+    if(!recentURIs.contains(uri)) {
```
Comment (adamantivm) [inline android_10/src/org/ros/android/MasterChooser.java]: Another formatting nit: space between `if` and open parenthesis (and `for`, etc.).

**Review by adamantivm — APPROVED** [review]
@stratomda thanks for the updates. I included a few more formatting / minor naming change requests.
Another question too: how would the user delete recent URIs? I would be OK with merging without having that ability, but then I'd appreciate creating an issue to remind us to add that capability somehow later.

File: android_10/src/org/ros/android/MasterChooser.java (hunk 4 above)
Comment (stratomda) [reply android_10/src/org/ros/android/MasterChooser.java]: Agree with you. Forgot to add in latest commit.

File: android_10/src/org/ros/android/MasterChooser.java (hunk 5 above)
Comment (stratomda) [reply android_10/src/org/ros/android/MasterChooser.java]: I'll address this. Didn't see it.

**Comment by stratomda** [comment]
Lets create an issue and we can hopefully add an option to the MasterChooserSettings to clear recent Master URIs. Does that sound like a plan? Otherwise in advanced options I could add an option to remove them as well.

**Comment by adamantivm** [comment]
@stratomda I would be perfectly OK with your proposal of an issue. Would you mind going ahead and creating it? Or alternatively updating the issue you and @jubeira have been using to track desired updates to the Master Chooser. Either would be fine.

**Pull request merged**